vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,797 @@
+            let visibleLineNumbers = Set [firstVisibleLineNumber .. lastVisibleLineNumber]
+            let nonVisibleLineNumbers = Set.difference recentVisibleLineNumbers visibleLineNumbers
+            let newVisibleLineNumbers = Set.difference visibleLineNumbers recentVisibleLineNumbers
+        
+            logInfof "nonVisibleLineNumbers = %A\newVisibleLineNumbers %A" nonVisibleLineNumbers newVisibleLineNumbers
+            if nonVisibleLineNumbers.Count > 0 || newVisibleLineNumbers.Count > 0 then
+                let snapshot = buffer.CurrentSnapshot
+                let view = view.Value
+                for lineNumber in nonVisibleLineNumbers do
+                    if lineNumber > 0 && lineNumber < snapshot.LineCount then
+                        try
+                            let line = 
+                                let l = e.NewSnapshot.GetLineFromLineNumber(lineNumber)
+                                view.GetTextViewLineContainingBufferPosition(l.Start)
+                            let tags = line.GetAdornmentTags(self)
+                            match tags |> Seq.tryHead with
+                            | None -> ()
+                            | Some tag ->
+                                let uis = tag :?> StackPanel seq
+                                for ui in uis do
+                                    if ui |> isNull |> not then
+                                        ui.Visibility <- Visibility.Hidden
+                        with e -> logErrorf "Error in non visible lines iteration %A" e
+                for lineNumber in newVisibleLineNumbers do
+                    try
+                        let line = 
+                                let l = e.NewSnapshot.GetLineFromLineNumber(lineNumber)
+                                view.GetTextViewLineContainingBufferPosition(l.Start)
+                        let tags = line.GetAdornmentTags(self)
+                        match tags |> Seq.tryHead with
+                        | None -> ()
+                        | Some tag ->
+                            let uis = tag :?> StackPanel seq
+                            for ui in uis do
+                                if ui |> isNull |> not then
+                                    ui.Visibility <- Visibility.Visible
+                                    layoutUIElementOnLine view line.Extent ui
+                     with e -> logErrorf "Error in new visible lines iteration %A" e
+            // Save the new first and last visible lines for tracking
+            recentFirstVsblLineNmbr <- firstVisibleLineNumber
+            recentLastVsblLineNmbr <- lastVisibleLineNumber
+            // We can cancel existing stuff because the algorithm supports abortion without any data loss
+            layoutChangedCts.Cancel()
+            layoutChangedCts.Dispose()
+            layoutChangedCts <- new CancellationTokenSource()
+            asyncMaybe {
+                do! Async.SwitchToContext uiContext |> liftAsync
```

Comment:
Please add a comment as to why the UI context is needed here

Reply:
Thanks for the reminder, I completely need to add a warning to the code that it expects to run on the UI thread or UI context.